File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    dx = [-1, 1, 0, 0]
    dy = [0, 0, -1, 1]
    
    n = len(maps)
    m = len(maps[0])
    
    if n == 1 and m == 1:
        return 1;
    
    dist = [[-1] * m for _ in range(n)]
    
    q = deque([(0,0)])
    dist[0][0] = 1
    
    while q:
        x,y = q.popleft()
        
        for i in range(4):
            nx, ny = x + dx[i], y + dy[i]
            if 0 <= nx < n and 0 <= ny < m:
                
                if maps[nx][ny] == 1 and dist[nx][ny] == -1:
                    dist[nx][ny] = dist[x][y] + 1
                    q.append((nx,ny))
                    
                    if nx == n-1 and ny == m-1:
                        return dist[nx][ny]
                    
                    
    return -1

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include <bits/stdc++.h>
using namespace std;

int solution(vector<vector<int> > maps)
{
    int dx[] = {-1,1,0,0};
    int dy[] = {0,0,-1,1};
    
    int n = maps.size();
    int m = maps[0].size();
    
    if (n == 1 && m == 1)
        return 1;
    
    vector<vector<int>> dist(n,vector<int>(m,-1));
    
    queue<pair<int,int>> q;
    q.push({0,0});
    dist[0][0] = 1;
    
    while(!q.empty()){
        int x = q.front().first;
        int y = q.front().second;
        q.pop();
            
        for (int i = 0; i < 4; i++){
            int nx = x + dx[i];
            int ny = y + dy[i];
            
            if (0 <= nx && nx < n && 0 <= ny && ny < m){
                if(maps[nx][ny] == 1 && dist[nx][ny] == -1){
                    dist[nx][ny] = dist[x][y] + 1;
                    q.push({nx,ny});
                    
                    if(nx == n-1 && ny == m-1)
                        return dist[nx][ny];
                }
            }
        }
    }
    return -1;
}